syntax = "proto3";

package metadata;

service MetadataService {
  rpc GetMovieReleaseDates (IMDbRequest) returns (ReleaseDatesResponse) {}
  rpc GetMovieMetadata (IMDbRequest) returns (MetadataResponse) {}
  rpc GetEpisodeMetadata (IMDbRequest) returns (MetadataResponse) {}
  rpc GetShowMetadata (IMDbRequest) returns (MetadataResponse) {}
  rpc GetShowSeasons (IMDbRequest) returns (ShowSeasonsResponse) {}
  rpc TMDbToIMDb (TMDbRequest) returns (IMDbResponse) {}
  rpc BatchGetMetadata (BatchIMDbRequest) returns (BatchMetadataResponse) {}
}

message IMDbRequest {
  string imdb_id = 1;
}

message TMDbRequest {
  string tmdb_id = 1;
}

message MetadataResponse {
  map<string, string> metadata = 1;
  string source = 2;
}

message ReleaseDatesResponse {
  string release_dates = 1;
  string source = 2;
}

message SeasonsResponse {
  repeated Season seasons = 1;
  string source = 2;
}

message Season {
  int32 season_number = 1;
  int32 episode_count = 2;
}

message Episode {
  int32 episode_number = 1;
  string title = 2;
  string first_aired = 3;
  int32 runtime = 4;
}

message IMDbResponse {
  string imdb_id = 1;
  string source = 2;
}

message BatchIMDbRequest {
  repeated string imdb_ids = 1;
}

message BatchMetadataResponse {
  map<string, MetadataResponse> results = 1;
}

message ShowSeasonsResponse {
  repeated Season seasons = 1;
  string source = 2;
}

message SeasonInfo {
  int32 episode_count = 1;
  map<string, EpisodeInfo> episodes = 2;
}

message EpisodeInfo {
  string first_aired = 1;
  int32 runtime = 2;
  string title = 3;
}